cedar-lean/Cedar/Thm/Data/Map.lean

```diff
@@ -47,7 +84,85 @@ theorem wf_iff_sorted {α β} [LT α] [DecidableLT α] [StrictLT α] {m : Map α
     replace h := List.sortedBy_implies_canonicalize_eq h
     rw [WellFormed, toList, kvs, make, h]
 
-/-! ### contains and mem -/
+theorem if_wf_then_exists_make [LT α] [DecidableLT α] (m : Map α β) :
+  m.WellFormed → ∃ list, m = Map.make list
+:= by
+  intro h₁
+  exists m.kvs
+
+/--
+  In well-formed maps, if there are two pairs with the same key, then they have
+  the same value
+-/
+theorem key_maps_to_one_value [DecidableEq α] [LT α] [DecidableLT α] [StrictLT α] (k : α) (v₁ v₂ : β) (m : Map α β) :
+  m.WellFormed →
+  (k, v₁) ∈ m.kvs →
+  (k, v₂) ∈ m.kvs →
+  v₁ = v₂
+:= by
+  simp only [wf_iff_sorted, toList]
+  intro wf h₁ h₂
+  have h₃ := List.mem_of_sortedBy_unique wf h₁ h₂ (by simp)
+  injection h₃
+
+/--
+  `make` on equivalent lists of (k,v) pairs produces equal maps
+
+  (Organizationally belongs in the `make` section, but is needed before that)
+-/
+theorem make_eqv [LT α] [DecidableLT α] {xs ys : List (α × β)}:
+  xs ≡ ys →
+  Map.make xs = Map.make ys
+:= by
+  intro h; unfold make; simp
+  sorry
```

Comment:
The one `sorry` in this PR.  I will continue working on this, but help/advice is also appreciated.

Reply:
I believe I have a proof that this lemma actually is _not valid_.  I discovered this by noticing that the `Set` equivalent (`Set.make_make_eqv`) relies on `List.equiv_implies_canonical_eq`, and working out that `List.equiv_implies_canonical_eq` _does not hold_ for `Prod.fst` (see comments on `List.equiv_implies_canonical_eq` added in this PR).

Specifically, consider
```
  xs = [(1, false), (1, true)]
  ys = [(1, true), (1, false)]
```
Even though `xs ≡ ys`, after `Map.make`, in one map, `1` will map to `false`, and in the other, `1` will map to `true`.

Reply:
Now the question is, how many other lemmas that currently rely on `Map.make_eqv` can we save.

Reply:
I think `eq_iff_kvs_equiv` should still be true, it just needs a different proof that doesn't rely on `make_eqv`.  At a minimum, I don't think the counterexample above falsifies `eq_iff_kvs_equiv`.

Reply:
Found a proof of `eq_iff_kvs_equiv` that uses `List.sortedBy_equiv_implies_eq` instead.